DBZ-2456 selective snapshot implementation

KaushikIyer16 opened the pull request:
https://issues.redhat.com/browse/DBZ-2456

Hi @jpechane, @gunnarmorling as per your comments on #1766. I have re-modeled the change to include a property snapshot.table, which would allow for selectively taking the snapshot and is honored by all types of connectors. So Effectively only the subset of tables specified in the above configuration are considered for snapshots and are locked, post which streaming resumes. Could you please check and let me know if this suffices.

My apologies for creating a new PR as I had some problems in merging the refs from upstream.

KaushikIyer16 commented:
Hi, @gunnarmorling and @jpechane were you able to take a look at this change? Currently, I am maintaining a custom snapshotter to achieve the same, and it would be easier for me to not maintain the custom jar if this were to be part of the main connector.

gunnarmorling commented on debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:
```diff
@@ -308,6 +310,14 @@ public static BinaryHandlingMode parse(String value, String defaultValue) {
             .withDescription("The maximum number of records that should be loaded into memory while performing a snapshot")
             .withValidation(Field::isNonNegativeInteger);
 
+    public static final Field SNAPSHOT_MODE_TABLES = Field.create("snapshot.table")
```
This could need some streamlining with existing option names and structures. How about "table.snapshot.include.list" as a name? And making it a list of regexps, as is done for the other similar options?

KaushikIyer16 replied:
Hi @gunnarmorling, apologies, if I have missed something, in order to make a `Field` a list of regexps, would the following suffice?

```
public static final Field SNAPSHOT_MODE_TABLES = Field.create("snapshot.include.collection.list")
            .withDisplayName("Snapshot Mode include Data Collection")
            .withType(Type.LIST)
            .withWidth(Width.LONG)
            .withImportance(Importance.MEDIUM)
            .withValidation(Field::isListOfRegex)
            .withDescription(
                    "this setting must be set to specify a list of tables/collections 
whose snapshot must be taken on creating or restarting the connector.");
```

gunnarmorling replied:
Yes, that's good for the definition. See other options (e.g. "table.include.list") for how this is handled then (splitting, evaluating this as regexp, etc.).

gunnarmorling commented on debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:
```diff
@@ -308,6 +310,14 @@ public static BinaryHandlingMode parse(String value, String defaultValue) {
             .withValidation(Field::isNonNegativeInteger);
 
+    public static final Field SNAPSHOT_MODE_TABLES = Field.create("snapshot.table")
+            .withDisplayName("Snapshot Mode Custom Tables")
```
Needs updating

gunnarmorling commented on debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:
```diff
@@ -524,6 +535,12 @@ public boolean getSanitizeFieldNames() {
         }
     }
 
+    public Set<String> getSnapshotAllowedTables() {
```
How about `getTablesToBeSnapshotted()`?

gunnarmorling commented on debezium-core/src/main/java/io/debezium/pipeline/source/AbstractSnapshotChangeEventSource.java:
```diff
@@ -81,6 +84,18 @@ public SnapshotResult execute(ChangeEventSourceContext context) throws Interrupt
         }
     }
 
+    protected <T extends DataCollectionId> Set<T> determineAllowedDataCollectionsForSnapshot(final Set<T> allDataCollections) {
```
`determineDataCollectionsToBeSnapshotted()`.

gunnarmorling commented on debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/PostgresConnectorIT.java:
```diff
@@ -1157,6 +1157,57 @@ public void shouldAllowForCustomSnapshot() throws InterruptedException {
         VerifyRecord.isValidRead(s2recs.get(1), PK_FIELD, 2);
     }
 
+    @Test
```
Could you add the `@FixFor` annotation (here and elsewhere)?

KaushikIyer16 replied:
Sure, Will add them across all tests.

gunnarmorling commented on debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresSnapshotChangeEventSource.java:
```diff
@@ -94,7 +94,7 @@ protected void connectionCreated(RelationalSnapshotContext snapshotContext) thro
             throws SQLException, InterruptedException {
         final Duration lockTimeout = connectorConfig.snapshotLockTimeout();
-        final Optional<String> lockStatement = snapshotter.snapshotTableLockingStatement(lockTimeout, schema.tableIds());
+        final Optional<String> lockStatement = snapshotter.snapshotTableLockingStatement(lockTimeout, snapshotContext.capturedTables);
```
Wow, that's quite something. @jpechane, I think that's good, a bit hard to believe we locked indeed _all_ tables before? Thinking of it, wasn't there even a report related to this?

gunnarmorling commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java:
```diff
@@ -254,6 +255,9 @@ protected void execute() {
         Set<TableId> lockedTables = Collections.emptySet();
 
+        final Set<String> snapshotAllowedTables = context.getConnectorConfig().getSnapshotAllowedTables();
+        final Predicate<TableId> isAllowedForSnapshot = tableId -> snapshotAllowedTables.size() == 0
```
Interesting style. Is there a reason for making this a predicate instead of simply a private method which takes the table id?

gunnarmorling reviewed (commented):
A few comments inline. I think overall this is really good, thanks @KaushikIyer16!

@jpechane, your review will be welcomed, too. Thanks!

jpechane commented on debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/SnapshotReader.java:
```diff
@@ -404,7 +408,8 @@ protected void execute() {
                                     logger.info("\t '{}' is not added among known tables", id);
                                 }
-                                if (filters.tableFilter().test(id)) {
+                                //
```
Please remove the comment

KaushikIyer16 replied:
Sure, will re-check elsewhere for the same

jpechane commented on debezium-core/src/main/java/io/debezium/config/CommonConnectorConfig.java:
```diff
@@ -308,6 +310,14 @@ public static BinaryHandlingMode parse(String value, String defaultValue) {
             .withDescription("The maximum number of records that should be loaded into memory while performing a snapshot")
             .withValidation(Field::isNonNegativeInteger);
 
+    public static final Field SNAPSHOT_MODE_TABLES = Field.create("snapshot.table")
```
Can't we use different name like snapshot.data.collection to encompass both tables and MongoDB collections?

gunnarmorling replied:
Ah yes, very good point!

jpechane commented on debezium-core/src/main/java/io/debezium/pipeline/source/AbstractSnapshotChangeEventSource.java:
```diff
@@ -81,6 +84,18 @@ public SnapshotResult execute(ChangeEventSourceContext context) throws Interrupt
     }
 
+    protected <T extends DataCollectionId> Set<T> determineAllowedDataCollectionsForSnapshot(final Set<T> allDataCollections) {
+        final Set<String> snapshotAllowedTables = connectorConfig.getSnapshotAllowedTables();
```
Again, should use generic names like snapshotAllowedDataCollections

jpechane commented on debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbSnapshotChangeEventSource.java:
```diff
@@ -342,7 +343,7 @@ private void createDataEventsForReplicaSet(ChangeEventSourceContext sourceContex
         LOGGER.info("Beginning snapshot of '{}' at {}", rsName, rsOffsetContext.getOffset());
 
-        final List<CollectionId> collections = primaryClient.collections();
+        final Set<CollectionId> collections = determineAllowedDataCollectionsForSnapshot(primaryClient.collections());
```
There is a planned Jira for snapshotting collections based on the include list order. Could we keep this as a List?

KaushikIyer16 replied:
In methods like `io.debezium.relational.RelationalSnapshotChangeEventSource#determineCapturedTables` a Set is required. so i could one of two things:

- make `determineDataCollectionsToBeSnapshotted` to take a Collection<T> and return a `List<T>` and in the above method i can wrap it again in a set by doing `new HashSet<>(determineDataCollectionsToBeSnapshotted(getAllTableIds(ctx)))`

- what do you feel if we used a `LInkedHashSet` as it would maintain insertion order.

gunnarmorling replied:
@KaushikIyer16, sorry, I'm not quite following: how is this change for MongoDB related to `RelationalSnapshotChangeEventSource` (which only is used by the relational Debezium connectors)?

KaushikIyer16 replied:
the same method is being used to filter out the collection names in both relational and non-relational cases. in one case it deals with sets and the other with lists. hence, I settled down to Set

jpechane commented:
Hi @KaushikIyer16 thanks a lot for the contribution. I've left couple of comments - mainly related to method/variables/fields naming to be generic enough so they make sense even in non-relational context.
Tha only major question from me is the change from List to Set in case of MongoDB connector.

KaushikIyer16 commented:
Thank you @jpechane, @gunnarmorling, I will rename all variables to make it relational/non-relational agnostic. I have replied wherever I have some doubts on how to proceed, thanks a lot for the quick review

gunnarmorling commented:
Hey @KaushikIyer16, are you still planning to wrap up this one? It'll need a rebase in any case. I also still think the list -> set changes should be reverted. "snapshot.collection.include.list" sounds good to me. Let me know if you wanted to discuss any other details. Thanks.

KaushikIyer16 commented:
Hi @gunnarmorling, yes I am almost done with the changes, i needed only one clarification, with regards to making the Field a list of regexp(https://github.com/debezium/debezium/pull/1791#discussion_r488793142) Could you please check that, i can rebase and update PR ASAP

gunnarmorling commented:
Ah, sorry, had missed this one. Just replied there.

KaushikIyer16 commented:
Hi @gunnarmorling, I have updated the PR with the suggested changes

* All naming has been generalized.
* I have updated the Field, it is being treated as a regex in `AbstractSnapshotChangeEventSource#determineDataCollectionsToBeSnapshotted`.
* I have added the `@FixFor` annotation everywhere necessary.
* Reverted the Set to List in the MongoDB connector.

gunnarmorling commented:
@KaushikIyer16, could you rebase this one more time? There are some conflicts due to concurrent work on master since then. Thanks!

KaushikIyer16 commented:
Hi @gunnarmorling, I have rebased the branch, but there seems to be an issue with the Ci/CD. On my local, it is compiling successfully with version `1.3` but I can see that with version `1.4` on the build section over [here](https://github.com/debezium/debezium/runs/1193735773?check_suite_focus=true), it is failing to find the dependency. Can you help me out as to what has to be done?

jpechane commented:
@KaushikIyer16 This is solved now

KaushikIyer16 commented:
thank you @jpechane, @gunnarmorling. I ran the tests that failed on Travis CI on my local and it seems to be passing. is it possible to trigger a re-run without committing? I can't seem to find the option. I have added necessary documentation wherever I thought this property applies, so requesting for another round of review.

KaushikIyer16 commented:
@gunnarmorling, @jpechane I have rebased the current branch once again. Can you please take a look and let me know what is pending to get the Pull Request merged? Thanks.

KaushikIyer16 commented:
@gunnarmorling , @jpechane i have also run the test that has failed in the CI/CD, `SqlServerConnectorIT.excludeMultipleColumnsWhenCaptureInstanceExcludesSingleColumn` and it is running successfully on my local. could you please take a look again?

jpechane commented:
@KaushikIyer16 Applied, thanks a lot for the nice piece of work. Sorry for ti taking so long
